_posts/2020-05-18-mirror-kafka-eventhub.md

```diff
@@ -0,0 +1,225 @@
+The application can run even locally or be published on the Azure cloud.
+
+To try the entire pipeline, we just miss sending messages and see how they are mirrored to Event Hub triggering the above function to be executed.
+To do so, just use the `kafka-console-producer` command linme tool provided by every default Apache Kafka installation.
+Start a new pod in the Kubernetes cluster for hosting the producer and types a couple of JSON messages as follow.
+
+> They have to be JSON structured messages because the default Azure Functions deserializer expects this format
+
+```shell
+kubectl -n kafka run kafka-producer -ti --image=strimzi/kafka:0.17.0-kafka-2.4.0 --rm=true --restart=Never -- bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9092 --topic testeh
+If you don't see a command prompt, try pressing enter.
+>{ "id": 1, "message": "Hello from Strimzi Mirror Maker" }
+>{ "id": 2, "message": "Here another mirrored message" }
+>
+```
+
+On the Azure Functions application, the messages will be logged as following.
+
+```shell
+[09/05/2020 14:02:17] Executing 'Functions.EventHubTriggerJava1' (Reason='', Id=6c54f1df-6a3d-4acc-a89a-4078b2049ddc)
+[09/05/2020 14:02:22] Java Event Hub trigger function executed.
+[09/05/2020 14:02:22] Length:1
+[09/05/2020 14:02:22] {"id":1,"message":"Hello from Strimzi Mirror Maker"}
+[09/05/2020 14:02:23] Function "EventHubTriggerJava1" (Id: 6c54f1df-6a3d-4acc-a89a-4078b2049ddc) invoked by Java Worker
+[09/05/2020 14:02:23] Executed 'Functions.EventHubTriggerJava1' (Succeeded, Id=6c54f1df-6a3d-4acc-a89a-4078b2049ddc)
+[09/05/2020 14:02:51] Executing 'Functions.EventHubTriggerJava1' (Reason='', Id=8edf21e3-94b3-4f4f-b2a7-8e92a311f9f0)
+[09/05/2020 14:02:51] Java Event Hub trigger function executed.
+[09/05/2020 14:02:51] Length:1
+[09/05/2020 14:02:51] {"id":2,"message":"Here another mirrored message"}
+[09/05/2020 14:02:51] Function "EventHubTriggerJava1" (Id: 8edf21e3-94b3-4f4f-b2a7-8e92a311f9f0) invoked by Java Worker
+[09/05/2020 14:02:51] Executed 'Functions.EventHubTriggerJava1' (Succeeded, Id=8edf21e3-94b3-4f4f-b2a7-8e92a311f9f0)
+```
+
+### Conclusion
+
+Integrating applications running on different platforms and different clouds is becoming quite common nowadays and the hybrid cloud based use cases are growing fast.
+Back to the scenario showed in this blog post, the Apache Kafka cluster on Kubernetes could run on any cloud provider (Azure, Amazon, IBM, GCP) and thanks to the Strimzi operators, its deployment and management is fairly simple as well as mirroring data to other systems like Event Hub using Kafka Mirror Maker.
+
+I hope this post has persuaded you of the benefit of using Strimzi and Mirror Maker to do such an integration with Azure.
+Let us know what you are going to integrate!
```

Comment:
I think it would be interesting to compare this with other ways of doing the same thing (e.g. why are we using MM1 and not MM2? Is there a dedicated Kafka Connector for functions which could provide an alternative?)

Reply:
Well, MM2 seems to have incompatibility problem at broker protocol level and I would like to dig more into it to figure out, if possible, what they are.
Regarding comparing with other stuff, is it something that should go just in the conclusion section? It seems it could deserve a complete blog post. Or ... what you mean? :-)

Reply:
@tombentley anything to add here?

Reply:
I think you need to say _something_, because "why did you chose to do it that way" is an obvious question which springs to mind (at least do this reader). It's more obvious given that you're using MM1 and all the momentum is behind MM2, so it seems like a peculiar choice to anyone who's aware of what's going on. If you wanted to commit to blogging about the MM2 alternative then you could say something about the future post, but I can totally understand if you didn't want to commit to that.

Reply:
Instead of in the conclusions could you add a note to the EH section (line 72) about the incompatibility? And link to an issue if one exists?

"Note: Whilst MirrorMaker2 is gaining traction in the Kafka ecosystem, there appear to be some protocol compatibility issues with the Azure EventHubs protocol head. Therefore this post uses MirrorMaker1 whilst this is being investigated."

Reply:
I mentioned that in the EH section as @sjwoodman suggested leaving the conclusion as it is.